scanner/src/main/kotlin/experimental/PackageProvenanceResolver.kt

```diff
@@ -0,0 +1,135 @@
+import org.ossreviewtoolkit.utils.collectMessagesAsString
+import org.ossreviewtoolkit.utils.createOrtTempDir
+import org.ossreviewtoolkit.utils.log
+import org.ossreviewtoolkit.utils.showStackTrace
+
+/**
+ * The [PackageProvenanceResolver] provides a function to resolve the [Provenance] of a [Package].
+ */
+interface PackageProvenanceResolver {
+    /**
+     * Resolve the [Provenance] of [pkg] based on the provided [sourceCodeOriginPriority]. If no provenance can be
+     * resolved [UnknownProvenance] is returned.
+     */
+    fun resolveProvenance(pkg: Package, sourceCodeOriginPriority: List<SourceCodeOrigin>): Provenance
+}
+
+/**
+ * The default implementation of [PackageProvenanceResolver].
+ */
+class DefaultPackageProvenanceResolver : PackageProvenanceResolver {
+    /**
+     * Resolve the [Provenance] of [pkg] based on the provided [sourceCodeOriginPriority]. For source artifacts it is
+     * verified that the [RemoteArtifact] does exist. For a VCS it is verified that the revision exists. If the revision
+     * provided by the [package][pkg] metadata does not exist or is missing, the function tries to guess the tag based
+     * on the name and version of the [package][pkg].
+     */
+    override fun resolveProvenance(pkg: Package, sourceCodeOriginPriority: List<SourceCodeOrigin>): Provenance {
+        sourceCodeOriginPriority.distinct().forEach { sourceCodeOrigin ->
+            kotlin.runCatching {
+                when (sourceCodeOrigin) {
+                    SourceCodeOrigin.ARTIFACT -> {
+                        if (pkg.sourceArtifact != RemoteArtifact.EMPTY) {
+                            return resolveSourceArtifact(pkg.sourceArtifact)
+                        }
+                    }
+
+                    SourceCodeOrigin.VCS -> {
+                        if (pkg.vcsProcessed != VcsInfo.EMPTY) {
+                            return resolveVcs(pkg)
+                        }
+                    }
+                }
+            }.onFailure {
+                it.showStackTrace()
+
+                log.info {
+                    "Could not resolve $sourceCodeOrigin for ${pkg.id.toCoordinates()}: ${it.collectMessagesAsString()}"
+                }
+            }
+        }
+
+        return UnknownProvenance
+    }
+
+    private fun resolveSourceArtifact(sourceArtifact: RemoteArtifact): ArtifactProvenance {
+        val request = Request.Builder().head().url(sourceArtifact.url).build()
+        OkHttpClientHelper.execute(request).use { response ->
+            if (response.code == HttpURLConnection.HTTP_OK) {
+                return ArtifactProvenance(sourceArtifact)
+            }
+
+            throw IOException("Could not verify existence of source artifact at ${sourceArtifact.url}.")
+        }
+    }
+
+    private fun resolveVcs(pkg: Package): RepositoryProvenance {
+        // TODO: Is it right to always determine type from URL?
+        val vcs = VersionControlSystem.forUrl(pkg.vcsProcessed.url)
```

Comment:
Maybe do `?: throw IOException("Could not determine VCS type for URL ${pkg.vcsProcessed}.")` here to avoid nesting below?